source/utilities.tex

```diff
@@ -16959,6 +17011,49 @@
 supported\iref{basic.align}.
 
 \indexlibrary{\idxcode{common_type}}%
+\pnum
+Let:
+\begin{itemize}
+\item \tcode{CREF(A)} be \tcode{add_lvalue_reference_t<const remove_reference_t<A>{}>},
+\item \tcode{XREF(A)} denote a unary class template \tcode{T} such that
+  \tcode{T<U>} denotes the same type as \tcode{U} with the addition
+  of \tcode{A}'s cv and reference qualifiers, for a type \tcode{U} such
+  that \tcode{is_same_v<U, remove_cvref_t<U>>} is \tcode{true},
+\item \tcode{COPYCV(FROM, TO)} be an alias for type \tcode{TO} with the addition of
+  \tcode{FROM}'s top-level cv-qualifiers.
+  \begin{example}
+    \tcode{COPYCV(const int, volatile short)} is an alias for
+    \tcode{const volatile short}.
+  \end{example}
+\item \tcode{COND_RES(X, Y)} be
+  \tcode{decltype(false ? declval<X(\&)()>()() : declval<Y(\&)()>()())}.
```

Comment:
Why are we creating two invented function types here, just to call them and look at their result values? What's the net difference vs. omitting the indirection via the function references? What's the net difference vs. `common_type_t<X,Y>`?

Reply:
This is a truly disgusting hack to avoid `declval`'s `add_rvalue_reference_t` on the return type. `decltype(declval<X>())` is `add_rvalue_reference_t<X>`, `decltype(declval<X(&)()>()()` is exactly `X`. We discussed at length in LWG that this is abominable, and that no one had an idea for an improvement.

Reply:
... and `common_type_t<X,Y>` doesn't work because?

Reply:
For cppreference, I just invented a `declval` substitute that returns `T` instead and used that, FWIW.

(`decltype(declval<X(&)()>()())` is also not exactly `X`; it's subject to the usual prvalue cv-qualification adjustment rules and the rvalue-reference-to-function-is-lvalue rule.)

Reply:
I'd hate to define yet another exposition-only helper here that's only going to be used in this one place. Do we agree that `template <class T> T __val() noexcept; // \expos` ... `decltype(false ? __val<X>() : __val<Y>())` is an improvement?

Reply:
> I'd hate to define yet another exposition-only helper here

I thought this discussion was familiar: https://github.com/ericniebler/stl2/issues/338#issuecomment-309616140.

Reply:
Maybe we shouldn't change anything here and go with the formulation in the paper. These nuances are important, but the code snippets for the meta-functions are way too obtuse to communicate that effectively. Can we maybe add a few notes or examples explaining what the intent is (in particular involving subtleties of prvalues vs. xvalues)? (As a matter of principle, I'm wondering whether some prose relying on the core language mechanics of the ?: operator or "common type" would be shorter and easier to understand. Not now.)